lib/identity_cache/cached/association.rb

```diff
@@ -39,6 +39,42 @@ def inverse_name
         end
       end
 
+      def fetch(records)
+        return if fetched?(records)
+
+        klass = reflection.active_record
+        cached_associations = klass.send(:embedded_associations)
+
+        return if cached_associations.empty?
+
+        return unless klass.primary_cache_index_enabled
+
+        cached_records_by_id = klass.fetch_multi(records.map(&:id)).index_by(&:id)
+
+        cached_associations.each_value do |cached_association|
+          records.each do |record|
+            next unless (cached_record = cached_records_by_id[record.id])
+            case cached_association
+            when Cached::Reference::HasMany
+              ids = cached_record.public_send(cached_association.cached_ids_name)
+              record.instance_variable_set(cached_association.ids_variable_name, ids)
+            when Cached::Reference::HasOne
+              id = cached_record.public_send(cached_association.cached_id_name)
+              record.instance_variable_set(cached_association.id_variable_name, id)
+            else
+              records = cached_record.public_send(cached_association.cached_accessor_name)
+              record.instance_variable_set(cached_association.records_variable_name, records)
+            end
+          end
+        end
```

Comment:
The advantage of moving the fetching to the cached association class is that we could avoid a lot of type checking that is still being done here.  Instead, the type specific logic can be moved into the derived cached association class that actually needs it.

For instance, none of this code makes sense for a cached belongs to association.